docs/src/lib/metadata.md

```diff
@@ -0,0 +1,294 @@
+# Metadata
+
+## Design of metadata support
+
+DataFrames.jl allows you to store and retrieve metadata on table and column
+level. This is supported using the functions defined by DataAPI.jl interface:
+
+* for table level metadata: [`metadata`](@ref), [`metadatakeys`](@ref),
+  [`metadata!`](@ref), [`deletemetadata!`](@ref), [`emptymetadata!`](@ref);
+* for column level metatadata: [`colmetadata`](@ref), [`colmetadatakeys`](@ref),
+  [`colmetadata!`](@ref), [`deletecolmetadata!`](@ref), [`emptycolmetadata!`](@ref).
+
+Assume that we work with a data frame-like object `df` that has a column `col`
+(referred to either via a `Symbol`, a string or an integer index).
+
+Table level metadata has a form of key-value pairs that are attached to `df`.
+Table level metadata has a form of key-value pairs that are attached to
+a specific column `col` of `df` data frame.
+
+Additionally each metadata key-value pair has a style information attached to
+it.
+In DataFrames.jl the metadata style influences how metadata is propagated when
+`df` is transformed. The following metadata styles are supported:
+
+* `:none`: metadata having this style is considered to be attached to a concrete
+  state of `df`; this means that any operation on this data frame,
+  invalidates such metadata and it is dropped in the result of such operation
+  (note that this happens even if the operation eventually does not change
+  the data frame; the rule is that calling a function that might alter a data
+  frame drops such metadata; in this way it is possible to statically determine
+  if metadata of styles other than `:note` is dropped after a function call);
+  the only exceptions that keep non-:note metadata are (the reason is that these
+  operations are specifically designed to create an identical copy of the source
+  data frame):
+    - [`DataFrame`](@ref) constructor;
+    - [`copy`](@ref) of a data frame;
+* `:note`: metadata having this style is considered to be an annotation of
+  a table or a column that should be propagated under transformations
+  (exact propagation rules of such metadata are described below);
+* all other metadata styles are allowed but they are currently treated as having
+  `:none` style (this might change in the future if other standard metadata
+  styles are defined).
+
+All DataAPI.jl functions work with [`DataFrame`](@ref),
+[`SubDataFrame`](@ref), [`DataFrameRow`](@ref)
+objects, and objects returned by [`eachrow`](@ref), and [`eachcol`](@ref)
+functions. In this section collectively these objects will be called
+*data frame-like*, and follow the rules:
+
+* objects returned by
+  [`eachrow`](@ref), and [`eachcol`](@ref) functions have the same metadata
+  as metadata of their parent `AbstractDataFrame`;
+* [`SubDataFrame`](@ref) and [`DataFrameRow`](@ref) have only metadata from
+  their parent `DataFrame` that has `:note` style.
```

Comment:
This is a crucial element of the whole design. It should be equivalent if you do:
```
df[rows, cols]
```
and
```
@view df[rows, cols]
```

Note that since `df[rows, cols]` materializes a new data frame only `:note` metadata is kept. So the same must happen for `@view df[rows, cols]`. The reason is that from user's perspective the fact if we use a view or materialize it should be transparent.

For example if you have column-level metadata that keeps column mean then it is invalidated by `@view df[rows, cols]` in the same way as it is invalidated by `df[rows, cols]`.

Reply:
Makes sense. That requires more code but it's indeed safer.